rust-runtime/aws-smithy-types/src/timeout.rs

```diff
@@ -0,0 +1,356 @@
+    pub fn read_timeout(&self) -> Option<Duration> {
+        self.read_timeout
+    }
+
+    /// A limit on the amount of time it takes for the first byte to be sent over an established,
+    /// open connection and when the last byte is received from the service for a single attempt.
+    /// If you want to set a timeout for an entire request including retry attempts,
+    /// use [`TimeoutConfig::api_call_timeout`] instead.
+    pub fn api_call_attempt_timeout(&self) -> Option<Duration> {
+        self.api_call_attempt_timeout
+    }
+
+    /// A limit on the amount of time it takes for request to complete. A single request may be
+    /// comprised of several attemps depending on an app's [`RetryConfig`](super::retry::RetryConfig). If you want
+    /// to control timeouts for a single attempt, use [`TimeoutConfig::api_call_attempt_timeout`].
+    pub fn api_call_timeout(&self) -> Option<Duration> {
+        self.api_call_timeout
+    }
+
+    /// Generate a human-readable list of the timeouts that are currently set. This is used for
+    /// logging and error reporting.
+    pub fn list_of_set_timeouts(&self) -> Vec<&'static str> {
+        let mut vec = Vec::new();
+        if self.api_call_timeout.is_some() {
+            vec.push("api call")
+        }
+        if self.api_call_attempt_timeout.is_some() {
+            vec.push("api call attempt")
+        }
+        if self.tls_negotiation_timeout.is_some() {
+            vec.push("TLS negotiation")
+        }
+        if self.connect_timeout.is_some() {
+            vec.push("connect")
+        }
+        if self.read_timeout.is_some() {
+            vec.push("read")
+        }
+
+        vec
+    }
+
+    /// Consume a `TimeoutConfig` to create a new one, setting the connect timeout
+    pub fn with_connect_timeout(mut self, timeout: Duration) -> Self {
+        self.connect_timeout = Some(timeout);
+        self
+    }
+
+    /// Consume a `TimeoutConfig` to create a new one, setting the TLS negotiation timeout
+    pub fn with_tls_negotiation_timeout(mut self, timeout: Duration) -> Self {
+        self.tls_negotiation_timeout = Some(timeout);
+        self
+    }
+
+    /// Consume a `TimeoutConfig` to create a new one, setting the read timeout
+    pub fn with_read_timeout(mut self, timeout: Duration) -> Self {
+        self.read_timeout = Some(timeout);
+        self
+    }
+
+    /// Consume a `TimeoutConfig` to create a new one, setting the API call attempt timeout
+    pub fn with_api_call_attempt_timeout(mut self, timeout: Duration) -> Self {
+        self.api_call_attempt_timeout = Some(timeout);
+        self
+    }
+
+    /// Consume a `TimeoutConfig` to create a new one, setting the API call timeout
+    pub fn with_api_call_timeout(mut self, timeout: Duration) -> Self {
+        self.api_call_timeout = Some(timeout);
+        self
+    }
+}
+
+/// A builder for [`TimeoutConfig`]s
+#[derive(Default, Debug, Clone, PartialEq)]
+pub struct TimeoutConfigBuilder {
```

Comment:
do we actually need this given that you can have a timeout config that's all empty? I know we have this pattern in other places, but it's worth thinking about

Reply:
I was asking myself the same thing in a comment in a previous version. I wasn't sure how important it was to uphold convention here. The same probably goes for the `RetryConfig`